### ISSUE TITLE
Nice paper - any way to do this for regression tasks?

### ISSUE THREAD

**Issue by Laksh1997** [wiseodd/last_layer_laplace#4]
Thanks for an awesome paper! I was wondering (and still am wondering...) about a way to apply this to regression tasks.

Have you all considered this?

**Comment by Laksh1997** [comment]
I was thinking estimate the hessian and then do a MC integral

**Comment by Laksh1997** [comment]
I've been using this function which works well on the blob dataset:

```python
@torch.no_grad()
def predict_mc(x, Sigma, num_samples=500):
  phi = model.feature_map(x).numpy() # Feature vector of x

  import torch.distributions as dist
  mvn = dist.MultivariateNormal(loc=W.data.squeeze(), covariance_matrix=torch.from_numpy(Sigma))
  sampled_weights = mvn.sample((num_samples, ))
  phi = torch.from_numpy(phi)

  logits = phi @ sampled_weights.T # 2500 x N_samples
  py = logits.sigmoid() # 2500 x N_samples

  mean_py = py.mean(1) # 2500 x N_samples

  var = (py - mean_py.unsqueeze(1)) ** 2 # 2500 x N_samples
  var = var.mean(1) # 2500, 

  confidence_estimate = 1 - var ** 0.5

  return mean_py, var, confidence_estimate
```

and it would transfer to regression easily.

However, when I increase the model dimension to 256, 500 samples from the multivariate normal no longer cuts it. Even 50k samples isn't ideal. And 500k blows ram!

**Comment by Laksh1997** [comment]
On second thoughts, I realised that the integral without the sigmoid is analytically tractable - the mean is just weight . data, the variance is data.T dot Covariance dot data

**Comment by wiseodd** [comment]
Yes, that's true: If `w ~ N(w | m, S)` and `f(x) = w^T x`, then `f(x) ~ N(f(x) | m^T x, x^T S x)`, so you don't need to sample. Also, in case you really need to sample (e.g. in classification), it's better to keep only the running average instead of the whole samples.
